example/advection/redefine_parthenon_defaults.cpp

```diff
@@ -48,4 +78,129 @@ void ParthenonManager::SetFillDerivedFunctions() {
+    const auto &profile = pkg->Param<int>("profile");
+
+    // TODO(pgrete) needs to be a reduction when using parallel_for
+    for (int k = kl; k <= ku; ++k) {
+      for (int j = jl; j <= ju; ++j) {
+        for (int i = il; i <= iu; ++i) {
+          Real ref_val;
+          if (profile == 0) { // wave
+            Real x =
+                cos_a2 * (pmb->coords.x1v(i) * cos_a3 + pmb->coords.x2v(j) * sin_a3) +
+                pmb->coords.x3v(k) * sin_a2;
+            Real sn = std::sin(k_par * x);
+            ref_val = 1.0 + amp * sn * vel;
+          } else if (profile == 1) { // smooth gaussian
+            Real rsq = pmb->coords.x1v(i) * pmb->coords.x1v(i) +
+                       pmb->coords.x2v(j) * pmb->coords.x2v(j) +
+                       pmb->coords.x3v(k) * pmb->coords.x3v(k);
+            ref_val = 1. + amp * exp(-100.0 * rsq);
+          } else if (profile == 2) { // hard_sphere
+            Real rsq = pmb->coords.x1v(i) * pmb->coords.x1v(i) +
+                       pmb->coords.x2v(j) * pmb->coords.x2v(j) +
+                       pmb->coords.x3v(k) * pmb->coords.x3v(k);
+            ref_val = (rsq < 0.15 * 0.15 ? 1.0 : 0.0);
+          } else {
+            ref_val = 1e9; // use an artifically large error
+          }
+
+          // Weight l1 error by cell volume
+          Real vol = pmb->coords.Volume(k, j, i);
+
+          l1_err += std::abs(ref_val - q(k, j, i)) * vol;
+          max_err = std::max(static_cast<Real>(std::abs(ref_val - q(k, j, i))), max_err);
+        }
+      }
+    }
+    pmb = pmb->next;
+  }
+
+  Real max_max_over_l1 = 0.0;
+
+#ifdef MPI_PARALLEL
+  if (Globals::my_rank == 0) {
+    MPI_Reduce(MPI_IN_PLACE, &l1_err, 1, MPI_ATHENA_REAL, MPI_SUM, 0, MPI_COMM_WORLD);
+    MPI_Reduce(MPI_IN_PLACE, &max_err, 1, MPI_ATHENA_REAL, MPI_MAX, 0, MPI_COMM_WORLD);
+  } else {
+    MPI_Reduce(&l1_err, &l1_err, 1, MPI_ATHENA_REAL, MPI_SUM, 0, MPI_COMM_WORLD);
+    MPI_Reduce(&max_err, &max_err, 1, MPI_ATHENA_REAL, MPI_MAX, 0, MPI_COMM_WORLD);
+  }
+#endif
+
+  // only the root process outputs the data
+  if (Globals::my_rank == 0) {
+    // normalize errors by number of cells
+    Real vol = (mesh_size.x1max - mesh_size.x1min) * (mesh_size.x2max - mesh_size.x2min) *
+               (mesh_size.x3max - mesh_size.x3min);
+    l1_err /= vol;
+    // compute rms error
+    max_max_over_l1 = std::max(max_max_over_l1, (max_err / l1_err));
+
+    // open output file and write out errors
+    std::string fname;
+    fname.assign("advection-errors.dat");
+    std::stringstream msg;
+    FILE *pfile;
+
+    // The file exists -- reopen the file in append mode
+    if ((pfile = std::fopen(fname.c_str(), "r")) != nullptr) {
+      if ((pfile = std::freopen(fname.c_str(), "a", pfile)) == nullptr) {
+        msg << "### FATAL ERROR in function Mesh::UserWorkAfterLoop" << std::endl
+            << "Error output file could not be opened" << std::endl;
+        PARTHENON_FAIL(msg.str());
+      }
```

Comment:
I wanted to get your opinion on this https://github.com/lanl/parthenon/issues/119.

Reply:
Just left a comment there. I'm not sure completely understand. Is #119 the same as #162 ?